internal/controllers/machineset/machineset_controller.go

```diff
@@ -239,17 +239,6 @@ func (r *Reconciler) reconcile(ctx context.Context, cluster *clusterv1.Cluster,
 		}))
 	}
 
-	// Make sure to reconcile the external infrastructure reference.
-	if err := reconcileExternalTemplateReference(ctx, r.Client, cluster, &machineSet.Spec.Template.Spec.InfrastructureRef); err != nil {
```

Comment:
The reason why we have reconcile external at the beginning of the reconcile loop is that it is important to have ownerRefs as soon as possible and also to ensure they are always in place.

If we move this `reconcileExternalTemplateReference` to where we are handling scale up, a side effect for this is that e.g. after move / after someone or something will remove owner reference, those won't be re-created until we scale up again.

/test pull-cluster-api-e2e-main
(should double check this assumption when we test ownerRef resilience)

Might be we can achieve the same goal you are looking for with this PR by keeping the call to `reconcileExternalTemplateReference`, but tolerating IsNotFound error when scaling down (len(machines) - int(*(ms.Spec.Replicas)) > 0)?

Reply:
Might be, this check should be smarted, and tolerate the error only if this isn't the current MS

Reply:
fixed, thank you!